### DISCUSSION TITLE
Add search possibility to Media Browser

### DISCUSSION THREAD

**Discussion by micha91** [home-assistant/architecture#677] [General] [answered]
# Context
Some integrations such as Yamaha MusicCast, Kodi, Plex or Spotify could offer filter/search capabilities in the media browser. This would enable the user to search for specific items instead of browsing around for a while.
# Proposal
We could add additional `can_search` and `search_string` attributes, which default to False/None, to the BrowseMedia class. If `can_search` is set, we could display a text field and a search button in the frontend on the top of the current view. The text field should be filled out with the value of the `search_string` attribute. When the user entered a search string in the frontend and hits the search button, the `browse_media` service should be called. This search string could be added to the `browse_media` service call as an additional `search_string` parameter. In the implementation of `browse_media` the search operation is applied and the BrowseMedia object is returned as usual, but with the applied search string set in the newly introduced `search_string` attribute.
In my opinion the search operation should always be performed using a function of the API, which is consumed by the integration, and *not* inside of Home Assistant.
In the UI it could look like this for directories, which set `can_search` to true:
![image](https://user-images.githubusercontent.com/20923302/145726002-f45bdda9-d875-4db1-abb7-0a8c9b620ebe.png)

**Comment by balloob** [comment]
Sounds like a good idea.

**Comment by allenporter** [comment]
Would we use something like this for filtering event types in the `nest` media player?  A search isn't exactly ideal (compared to a filter) but it isn't the worst way this could work. This would require local filtering, going against the search via api recommendations. Here is an example now:

<img width="977" alt="Screen Shot 2021-12-18 at 11 13 31 AM" src="https://user-images.githubusercontent.com/6026418/146653171-85267784-80cb-4f03-9d04-4d53097387c6.png">

An alternative for this use case could be to expose additional virtual directories as event type filters

**Reply by balloob** [reply]
You can test a users query against the event types (sound, motion etc)

**Comment by OnFreund** [comment]
I missed this discussion and opened https://github.com/home-assistant/architecture/discussions/1038.

I really think this has become even more relevant with the recent advancement in assistants.

**Comment by OnFreund** [comment, marked as answer]
Trying to resurrect this, I'm thinking through the backend implementation. I think search can be added to both media players and media sources (in the same way that both can support browsing).
The common interface would be a search function:

```python
async def async_search_media(self, query: str, allowed_classes: list[MediaClass]) -> list[BrowseMediaSource]
```

The return value would be a list corresponding to the media classes available in the results, with the children of each `BrowseMediaSource` representing the actual items. In the future we can also consider adding a limit to the number of results, and an offset/page to allow retrieving additional results.

Media players would signal their support using a newly added feature flag - `MediaPlayerEntityFeature.SEARCH_MEDIA`.
For media sources we'll add a new `MediaSourceEntityFeature` type with both `BROWSE_MEDIA` and `SEARCH_MEDIA` constants.

What do you think?

**Reply by balloob** [reply]
I like this idea. I wonder if it's allowed_classes or that we should work like we do for browse_media, which is to pass in a filter function?

**Reply by OnFreund** [reply]
I'm not familiar with the filter function, but the main question is whether we can filter at the source. Unlike browse, with search it's important that the filter is passed to the underlying service to avoid the extra bandwidth (think of something like spotify returning lots of songs instead of a single artist, if you're searching for one).

**Reply by OnFreund** [reply]
Ok, found the [filter function](https://github.com/home-assistant/core/blob/1d62f0e380c9f9bb35a9c5d30293a05526dcbb87/homeassistant/components/media_source/__init__.py#L128) - it's indeed client side. I think that we should be able to limit the class of results at the service as well.

**Reply by emontnemery** [reply]
I did chat with @balloob about this, we suggest to rename `allowed_classes` to `allowed_content_types`.
Also, we should consider allowing asterisks, so `image/*` etc.

**Reply by OnFreund** [reply]
Thanks. A few challenges I see with content types:
1. The semantics aren't clear - if the user wants to search for either albums or playlists, it will be the same content type (e.g. m3u)
2. I think most 3rd party services aren't going to have this granularity when searching, so we might see a lot of repetition where integrations convert those content types to classes.
3. I also think users would prefer marking classes (e.g. "I want either songs or albums", rather than "I want either mp3s or m3us")

**Reply by OnFreund** [reply]
I think we can take inspiration from Music Assistant here:
![Screenshot 2024-10-03 at 5 51 12 PM](https://github.com/user-attachments/assets/c7c71b28-f10e-4023-aa52-c82dee37b112)

**Reply by balloob** [reply]
Yes, let's take inspiration from Music Assistant and use those media types.

**Reply by emontnemery** [reply]
To clarify, we should allow searching by (possibly a subset) of the `MediaClass` enum

**Reply by peteS-UK** [reply]
I agree that the enum is the right place to start - media type really reflects the location of the item in the browser hierarchy rather than the file format.  Maybe it should be the MediaType enum though, rather than MediaClass.  However, there needs to be an "Other" option as well as integrations might extend MediaType for things that are specific to the endpoint.  So, for example, in the squeezbox integration, we have a top level Apps type, but each app actually is its own type (e.g. Spotify, Qubuz, Tidal etc.) , or I also have a top level of Radios for radio plugins, and then each plugin is its own app.  I was going to use Favorites as an example as well, but that's so generic I think it should have its own type anyway - MediaType.FAVORITE = "favorite" should be defined I think - I'll add a PR for that one.

**Reply by OnFreund** [reply]
Looks like this was added to Media Player in https://github.com/home-assistant/core/pull/140321 and is available in 2025.5

**Reply by peteS-UK** [reply]
> Looks like this was added to Media Player in [home-assistant/core#140321](https://github.com/home-assistant/core/pull/140321) and is available in 2025.5

...Yes - partly.  The new PR adds a media_search action, and a new function in media player async_media_search that it uses.  In the future, this will give the groundwork for search in the UI in media browser, but that part is not done yet.  Now, media player integrations need to implement the MEDIA_SEARCH function.

**Reply by mill1000** [reply]
I would to like to see search functionality for UPnP media servers + players. 

Would implementing this via the MediaSource class be the correct approach? In other words, would that allow voice assistants to perform searches or are they only capable of performing a search against a media player?

Alternative, could we "bind" a media source to a player and then perform searches via the media player class?

**Comment by balloob** [comment]
Implemented in 2025.5
